### ISSUE TITLE
Minor Kerning Interpolation Issue.

### ISSUE THREAD

**Issue by VivaRado** [fonttools/fonttools#1411]
https://github.com/fonttools/fonttools/blob/6e1e0b73f1d82a0a3d1e6307164dddcadcf20df6/Lib/fontTools/varLib/merger.py#L677

Reversing cols when first values are None, solves minor kerning interpolation issue on compile.
Tested only for Group to Group based kerning pairs.

```
...
	...
		for cols in zip(*list(r.Class2Record for r in rows)):
			#
			if cols[0] == None:
				cols = cols[::-1]
			#
	...
...
```

**Comment by anthrotype** [comment]
you propose a "solution" but I still don't get whant the issue is

**Comment by VivaRado** [comment]
Simple problem im trying to solve here. I have set extreme values between kerning the regular and italic. Values like -200 and 200. And I am trying to generate that.

By providing a small pair set (250 Group to Group), modifying the italics and regular to have opposite exaggerated kerning values (reg -200 and ital 200), and testing  it until varlib generates the interpolation between those font instances, we experience inconsistently generated VadData items. 

If you compare the OTFs, from two separate generation occasions of the same sources, that where **successful** at interpolating the extreme values, you get very small differences in parts of the header where i think its just the date.

Whereas if you compare the generated OTFs of an occasion where one interpolates and the other doesn't, the files have significant differences in DIFF.

If we open the files as TTX we see some data missing in the VarData tag and later on the ClassRecord. If we replace that data in the TTX with the data from the successfully generated kerning interpolating font and convert back to a TTF, we got kerning interpolation working on the file that it didn't work.

I hope that is clear, I could send you some files so you can check if you want. I've been all over this piece of code varLib, fontTools, fontproject following the trace. and it leads me to this issue, which my opinion is that is a surface issue. During I have experienced other issues with class assignments and mixed kerning pair sets getting miss assigned, or even classes created without existing in the provided classes which might be a feaLib issue, which I'm also looking into.

To be clear this is not an easy to identify bug and one might think that all the generated values look ok but in reality they are not. that is if you put some exaggerated values for good measure and see if they actually get outputted as they should. On small kerning pair sets, the kerning interpolation succeeds more than if the kerning.fea is larger with more pairs, involving Group to Group, Group to Letter, letter to Group, and Letter to Letter. 

Thank you!

**Comment by anthrotype** [comment]
i'm sorry, I'm even more confused than before. Maybe somebody else can help you.

**Comment by VivaRado** [comment]
No problem, I'm helping myself. Just checkup on the PR

**Comment by VivaRado** [comment]
Hello! Great day to you all! Thank you for your time.
In the Advent repository I have included the files I'm using to test for example a Group to Group kerning list of length 400.

If you run the [designspace located here](https://github.com/VivaRado/Advent/blob/master/font_source/EFO/temp/font.designspace), and create a variable OTF. And [replace it here](https://github.com/VivaRado/Advent/blob/master/font_preview/assets/fonts/advent_pro_var_kern_comp.otf).

In the preview HTML, [located here](https://github.com/VivaRado/Advent/blob/master/font_preview/index.html), you should, when using the italics slider, see the "AV" pair. Start from a negative -292 and reach a positive 262.

If you compile it a few times, you will see that the results are different. The more pairs you give it the less possible it is to generate the required result. I think my group definitions look ok, FEA and PLIST files are clean. But if you notice anything not ok there I would like to know so I can fix it. If you have any suggestions on what is the most efficient way to create kerning for a variable font I would be glad to abide.

If you then open the successfully generate result - where "AV" pair goes from negative to positive exaggeration - and the one that is generated and fails to interpolate - as a converted TTX, you will notice that there are a lot of VarData / ClassRecord items missing in the one that failed to interpolate.

If you then replace that missing data from the successful to the one that failed. Seemingly since its from the same source it should be the same but its not. The failed one, as expected, after replacement, works and successfully interpolates the kerning between upright and regular.

And after a lot of tracing in the Varlib, Fonttools, Fealib etc... I assumed the issue lies within the locations where those lines of XML - dicts, are generated. That makes me believe that the issue happens in the merger.

Thank you for the help.

**Comment by behdad** [comment]
What's EFO?  How do I build this?  Please upload here a zip file with a designspace file and the ttf-interpolatable output from fontmake, that I can call `fonttools varLib` on and get the variable font.  Thanks.

**Comment by behdad** [comment]
Thanks.  Note that you can also upload zip files to github issues.

**Comment by behdad** [comment]
I can reproduce the issue.  If I keep running with python3, most times the output is 118840 bytes, but sometimes 118676 or 118716.  Doesn't happen with python2.  Digging more.

**Comment by behdad** [comment]
@anthrotype do you have ideas of what kind of things python3 can result in such nondeterminism?

**Comment by anthrotype** [comment]
It may be the hash randomisation, which is on by default in python >= 3.3. It’s designed to prevent attackers from maliciously triggering hash collisions.
We need to check all the sets or dictionaries and check we are not relying on any particular order of the collections (which changes at every run), but instead use a deterministic sorted order.
Try setting PYTHONHASHSEED environment variable to 0 to disable this feature and verify that the output doesn’t change at each run.

**Comment by VivaRado** [comment]
My guess so far is that it happens in ttLib/otTables:splitPairPos

**Comment by VivaRado** [comment]
Could this be it? in _ClassDef_invert

```

for s in sorted(ret[1:]):

instead of:

for s in ret[1:]:

```

**Comment by behdad** [comment]
I see the issue now.  Working on a fix.

**Comment by behdad** [comment]
Fixed.  Would be nice to add a test...  Should happen easily when there are more than one PairPosFormat2 subtables.

**Comment by VivaRado** [comment]
@behdad Well done! that's a big OOOF. Could you pinpoint the area so I can confirm please?

**Comment by VivaRado** [comment]
Hello, as i tested with the new fix that @behdad provided, the whole kerning pairs set:
```
{'GG': 3782, 'GL': 2048, 'LG': 2046, 'LL': 1089}
Total Pairs 8965
```
Was nicely compiling the the exaggerated values. But there is still something wrong.
Because i have to keep producing the font - and I cant really say or pinpoint the area of the issue for the inconsistency. Because there still is one. I can provide the deducted source - [you will find here](https://github.com/VivaRado/fontmake_deducted)

Where, even the slightest change made to the kerning plist - is successfully compiled.
After the update @behdad did, i adjusted the compile to include MVAR and GVAR.
I don't know what creates the kerning interpolation issue - might be some part where there is an optimization. But i know that the deducted gives me a result I want. As you are more knowledgeable in these libraries i trust you will be able to pinpoint the issue easier than me.

I just suggest a compile of [advent_for_test.zip](https://github.com/fonttools/fonttools/files/2701340/advent_for_test.zip)

with the deducted and the standard libraries. To compile with the deducted just point to fontmake/__main__

You will be able to see the differences again in the common culprits "AV".

Thank you! I will be able to check up on where there might be an issue as soon as i finish compiling the final Advent kerning.

**Comment by behdad** [comment]
I'm not sure what you are talking about.  Please file a new issue and give exact minimal instructions of how to reproduce the bug, what you get, and what you expect.

**Comment by VivaRado** [comment]
Ok, here is the preview build (sorry it took so long
[preview_build_2.zip](https://github.com/fonttools/fonttools/files/2701473/preview_build_2.zip)
): 

In the preview, there are two fonts:

 - Normal:
   - Here compiled with deducted in (index_normal.html):
      - We don't ask any exaggeration and "AVA", "AUA", "AWA" are ok.
   - Here compiled with standard in (index_normal_standard.html):
      - We don't ask any exaggeration, but "AVA", "AUA", "AWA" are NOT ok.
 - Exaggerated:
   - Here these values are exaggerated and compiled with deducted in (index_exaggerated_deducted.html):
      - All values exaggerate, but "AVA", "AUA", "AWA" are ok.
      - Group to Group:
         - thn (thin): pos @_A @_e1 200;
         - thn_it (thin italics): pos @_A @_e1 -234;
      - Group to Letter:
         - thn (thin): pos @_A AE 233;
         - thn_it (thin italics): pos @_A AE -252;
      - Letter to Group:
         - thn (thin): pos Hbar @_A1 -221;
         - thn_it (thin italics): pos Hbar @_A1 215;
      - Letter to Letter:
         - thn (thin): pos OE AE 220;
         - thn_it (thin italics): pos OE AE -211;
   - Here these same values compiled with standard in (index_exaggerated_standard.html):
      - All values exaggerate, but "AVA", "AUA", "AWA" are NOT ok.

In short: 
 - with the standard libraries, we get the values to exaggerate on **thin_it** and **thin** (Ae AÆ ŒÆ), but kerning in some other pairs gets affected.
 - with the deducted libraries, we get the values to exaggerate on **thin_it** and **thin** (Ae AÆ ŒÆ), and kerning in all other pairs is ok.

**Comment by behdad** [comment]
New issue please.
